src/market/directory.rs

```diff
@@ -315,32 +446,98 @@ pub fn start_directory_server(directory: Arc<DirectoryServer>) -> Result<(), Dir
 fn handle_client(
     stream: &mut TcpStream,
     directory: &Arc<DirectoryServer>,
+    wallet: &Wallet,
 ) -> Result<(), DirectoryServerError> {
-    let reader_stream = stream.try_clone()?;
-    let mut reader = BufReader::new(reader_stream);
-    let mut request_line = String::new();
-
-    reader.read_line(&mut request_line)?;
-    if request_line.starts_with("POST") {
-        let addr: String = request_line.replace("POST ", "").trim().to_string();
-        directory.addresses.write()?.insert(addr.clone());
-        log::info!("Got new maker address: {}", addr);
-    } else if request_line.starts_with("GET") {
-        log::info!("Taker pinged the directory server");
-        let response = directory
-            .addresses
-            .read()?
-            .iter()
-            .fold(String::new(), |acc, addr| acc + addr + "\n");
-        stream.write_all(response.as_bytes())?;
-        stream.flush()?;
+    let mut reader = BufReader::new(stream.try_clone()?);
+    let mut writer = BufWriter::new(stream);
+    let mut length_bytes = [0u8; 8];
+    reader.read_exact(&mut length_bytes)?;
+    let length = u64::from_be_bytes(length_bytes);
+
+    let mut buf = vec![0; length as usize];
+    reader.read_exact(&mut buf)?;
+    let value: Request = serde_cbor::de::from_reader(&buf[..])?;
+    match value {
+        Request::Post { metadata } => {
+            log::info!("Received new maker address: {}", &metadata.url);
+
+            let fidelity_redeem_script =
+                fidelity_redeemscript(&metadata.proof.bond.lock_time, &metadata.proof.bond.pubkey);
+            let address = Address::p2wsh(
+                fidelity_redeem_script.as_script(),
+                bitcoin::network::Network::Regtest,
+            );
+            let script_pubkey = &address.script_pubkey();
+
+            let txid = metadata.proof.bond.outpoint.txid;
+            let transaction = wallet.rpc.get_raw_transaction(&txid, None)?;
+            let tx_out = transaction.tx_out(0)?;
+            let redeem_script_pubkey = &tx_out.script_pubkey;
+
+            let cert_hash = metadata.proof.cert_hash;
+            let calculated_cert_hash = metadata.proof.bond.generate_cert_hash(&metadata.url);
+
+            let secp = Secp256k1::new();
+            let cert_message = Message::from_digest_slice(cert_hash.as_byte_array())?;
+
+            let ecdsa_verified = secp
+                .verify_ecdsa(
+                    &cert_message,
+                    &metadata.proof.cert_sig,
+                    &metadata.proof.bond.pubkey.inner,
+                )
+                .is_ok();
+
+            let current_height = wallet.rpc.get_block_count()?;
+            let lock_time_valid =
+                LockTime::from_height(current_height as u32)? < metadata.proof.bond.lock_time;
+
+            if script_pubkey == redeem_script_pubkey
+                && ecdsa_verified
+                && lock_time_valid
+                && cert_hash == calculated_cert_hash
```

Comment:
This is not a good idea from security standpoint. Ideally each check should be done sequentially, one after other. Starting from the least costly check ending with the most costly check. Merging them together can create potential attack like, someone keeps triggering costly secp verification, by just sending garbage tx.

You wanna put the checks one after another, only moving to the next check if the previous passes. Or else Error quickly.

Reply:
Just a pattern. Not a requirement. That also makes visualising the checks in the code easier.